glean-core/rlb/src/test.rs

```diff
@@ -1131,3 +1134,80 @@ fn test_boolean_get_num_errors() {
 
     assert_eq!(result, 0);
 }
+
+#[test]
+fn signaling_done() {
+    let _lock = lock_test();
+
+    // Define a fake uploader that reports back the submission URL
+    // using a crossbeam channel.
+    #[derive(Debug)]
+    pub struct FakeUploader {
+        barrier: Arc<Barrier>,
+        counter: Arc<Mutex<HashMap<ThreadId, u32>>>,
+    }
+    impl net::PingUploader for FakeUploader {
+        fn upload(
+            &self,
+            _url: String,
+            _body: Vec<u8>,
+            _headers: Vec<(String, String)>,
+        ) -> net::UploadResult {
+            let mut map = self.counter.lock().unwrap();
+            *map.entry(thread::current().id()).or_insert(0) += 1;
+
+            // Wait for the sync.
+            self.barrier.wait();
+
+            // Signal that this uploader thread is done.
+            net::UploadResult::done()
+        }
+    }
+
+    // Create a custom configuration to use a fake uploader.
+    let dir = tempfile::tempdir().unwrap();
+    let tmpname = dir.path().to_path_buf();
+
+    // We use a barrier to sync this test thread with the uploader thread.
+    let barrier = Arc::new(Barrier::new(2));
+    // We count how many times `upload` was invoked per thread.
+    let call_count = Arc::new(Mutex::default());
+
+    let cfg = Configuration {
+        data_path: tmpname,
+        application_id: GLOBAL_APPLICATION_ID.into(),
+        upload_enabled: true,
+        max_events: None,
+        delay_ping_lifetime_io: false,
+        server_endpoint: Some("invalid-test-host".into()),
+        uploader: Some(Box::new(FakeUploader {
+            barrier: Arc::clone(&barrier),
+            counter: Arc::clone(&call_count),
+        })),
+        use_core_mps: false,
+    };
+
+    let _t = new_glean(Some(cfg), true);
+
+    // Define a new ping and submit it.
+    const PING_NAME: &str = "test-ping";
+    let custom_ping = private::PingType::new(PING_NAME, true, true, vec![]);
+    custom_ping.submit(None);
+    custom_ping.submit(None);
+
+    // Sync up with the upload thread.
+    barrier.wait();
```

Comment:
Ah, and if the threading model changes (ie, if each `submit` starts getting its own thread) this test will fail by locking itself?

Reply:
yep

Reply:
well, in theory. *goes back to the code* why doesn't this fail now? oO

Reply:
ugh, turns out I'm still running into timing issues here. Let me once again rethink the test.